### PR TITLE
restructure add-items and db insertion with it

### PR DESCRIPTION
**Pull request by smadbe**
Probably the most important is to review `add-items.go`. 
There is still more general work to be done on this service (tests, permissions, non-dummy data, ...)

### CODE REVIEW

File: app/database/dataStore.go [hunk 1]
```diff
@@ -2,6 +2,8 @@ package database
 
 import (
 	"math/rand"
+
+	t "github.com/France-ioi/AlgoreaBackend/app/types"
```
Comment (slomek) [inline app/database/dataStore.go]: I would recommend not to alias this and just use `types`. In Go, package is a part of the name of whatever is exposed there.

File: app/database/dataStore.go [hunk 2]
```diff
@@ -49,8 +51,23 @@ func (s *DataStore) GroupAncestors() *GroupAncestorStore {
...
+// NewID generates a positive random int64 to be used as ID
+// !!! To be safe, the insertion should be be retried if the ID conflicts with an existing entry
+func (s *DataStore) NewID() int64 {
 	// gen a 63-bits number as we want unsigned number stored in a 64-bits signed DB attribute
-	// !!! to be safe, the insertion should be be retried if the ID conflicts with an existing entry
 	return rand.Int63()
 }
+
+// GenIDIfNotSet does check the given ID is set. If not, generate a (random) ID for it
+func (s *DataStore) GenIDIfNotSet(id *t.Int64) {
```
Comment (slomek) [inline app/database/dataStore.go]: Maybe `EnsureSetID`?

File: app/database/dataStore.go [hunk 3]
```diff
@@ -49,8 +51,23 @@ func (s *DataStore) GroupAncestors() *GroupAncestorStore {
...
+	if !id.Set {
+		*id = *t.NewInt64(s.NewID())
+	}
+}
+
+// InTransaction execute the given function in a transaction and commit
```
Comment (slomek) [inline app/database/dataStore.go]:
```suggestion
// InTransaction executes the given function in a transaction and commits
```

File: app/database/dataStore.go [hunk 4]
```diff
@@ -49,8 +51,23 @@ func (s *DataStore) GroupAncestors() *GroupAncestorStore {
...
+func (s *DataStore) InTransaction(txFunc func(*DataStore) error) error {
+	return s.db.inTransaction(func(db *DB) error {
+		return txFunc(&DataStore{db})
```
Comment (slomek) [inline app/database/dataStore.go]: Use names when assigning values to struct fields.

File: app/database/db.go [hunk 5]
```diff
@@ -17,11 +17,9 @@ type DB struct {
 	*gorm.DB
 }
 
-const dbStructTag string = "db"
-
-// DBConn connects to the database and test the connection
+// Open connects to the database and test the connection
```
Comment (slomek) [inline app/database/db.go]:
```suggestion
// Open connects to the database and tests the connection
```

File: app/database/db.go [hunk 6]
```diff
@@ -63,6 +61,7 @@ func (db *DB) inTransaction(txFunc func(*DB) error) (err error) {
 // insert reads fields from the data struct and insert the values which have been set
 // into the given table
 func (db *DB) insert(tableName string, data interface{}) error {
+
```
Comment (slomek) [inline app/database/db.go]: rm

File: app/api/items/add-item.go [hunk 7]
```diff
@@ -12,56 +12,96 @@ import (
...
 }
 
-// NewItemResponseData is what will be returned as data in case of success
-type NewItemResponseData struct {
-	ItemID int64 `json:"ID"`
+func (in *NewItemRequest) stringData(id int64) *database.ItemString {
+	return &database.ItemString{
+		ID:         *t.NewInt64(id),
+		ItemID:     in.ID.Int64,
+		LanguageID: in.Strings[0].LanguageID.Int64,
```
[the remark this answers is not in the document]
Comment (smadbe) [reply app/api/items/add-item.go]: Oh I thought it was done, thx. Fixed, tests added.

File: app/database/dataStore.go (hunk 4 above)
Comment (smadbe) [reply app/database/dataStore.go]: on external package it is important (required?). Here, I don't see how `&DataStore{db:db}` makes it more readable than `&DataStore{db}`

File: app/api/items/add-item.go [hunk 8]
```diff
@@ -12,56 +12,96 @@ import (
...
+
+func (srv *Service) insertItem(input *NewItemRequest) error {
+	srv.Store.GenIDIfNotSet(&input.ID.Int64)
+
+	return srv.Store.InTransaction(func(store *database.DataStore) error {
+		var err error
+		if err = store.Items().Insert(input.itemData()); err != nil {
+			return err
+		}
+		if err = store.GroupItems().Insert(input.groupItemData(store.NewID())); err != nil {
```
[the remark this answers is not in the document]
Comment (smadbe) [reply app/api/items/add-item.go]: to be discussed with the other styling stuffs

File: app/database/dataStore.go (hunk 4 above)
Comment (slomek) [reply app/database/dataStore.go]: There is a `go vet` warning: https://golang.org/cmd/vet/#hdr-Unkeyed_composite_literals

Comment (smadbe) [reply app/database/dataStore.go]: How do you get this warning? I do not have it locally, also it is not raised by golangci (which includes go vet) and `make lint` which includes go vet. Are you on Go 1.11 ?

Comment (slomek) [reply app/database/dataStore.go]: Hm, I used to get such warnings i VSCode and still do for primitive values, but here is doesn't show up indeed. Never mind then.

**Pull request merged**